### PR TITLE
[te][benchmark] Add more optimized versions of gemm

### PR DESCRIPTION
**Pull request by bertmaher**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48159 [te][benchmark] Add more optimized versions of gemm**
* #48158 [te] Make BUILD_TENSOREXPR_BENCHMARK a real CMake option

### CODE REVIEW

File: benchmarks/cpp/tensorexpr/bench_gemm.cpp [hunk 1]
```diff
@@ -127,6 +127,216 @@ BENCHMARK_DEFINE_F(Gemm, TensorExprTile32x32)(benchmark::State& state) {
...
+}
+
 BENCHMARK_REGISTER_F(Gemm, Torch)->Args({128, 128, 128});
 BENCHMARK_REGISTER_F(Gemm, TensorExprNoopt)->Args({128, 128, 128});
 BENCHMARK_REGISTER_F(Gemm, TensorExprTile32x32)->Args({128, 128, 128});
+BENCHMARK_REGISTER_F(Gemm, TensorExprTile4x16)->Args({128, 128, 128});
+BENCHMARK_REGISTER_F(Gemm, TensorExprTile4x16VecUnroll)->Args({128, 128, 128});
+BENCHMARK_REGISTER_F(Gemm, TensorExprTile4x16Cache)->Args({128, 128, 128});
+
+BENCHMARK_MAIN();
```
Comment (ZolotukhinM) [inline benchmarks/cpp/tensorexpr/bench_gemm.cpp:342]: Hm.. Why didn't we have this before?

File: benchmarks/cpp/tensorexpr/bench_gemm.cpp [hunk 2]
```diff
@@ -127,6 +127,216 @@ BENCHMARK_DEFINE_F(Gemm, TensorExprTile32x32)(benchmark::State& state) {
...
+    loop.reorderAxis(mi, k);
+  }
+  {
+    auto const& loops = loop.getLoopStmtsFor(CT);
+    loop.cacheAccesses(CT->buf(), "C_regs", loops[2]);
+  }
+
+  loop.prepareForCodegen();
+  te::Stmt* s = loop.root_stmt();
+  std::cerr << *s << "\n";
```
Comment (ZolotukhinM) [inline benchmarks/cpp/tensorexpr/bench_gemm.cpp:326]: Nit: leftover print.

File: benchmarks/cpp/tensorexpr/bench_gemm.cpp [hunk 3]
```diff
@@ -127,6 +127,216 @@ BENCHMARK_DEFINE_F(Gemm, TensorExprTile32x32)(benchmark::State& state) {
...
+    te::For* mi = loops[2];
+    te::For* k = loops[3];
+    loop.reorderAxis(mi, k);
+  }
+  {
+    auto const& loops = loop.getLoopStmtsFor(CT);
+    te::For* mi = loops[3];
+    te::For* ni = loops[4];
+    te::Stmt* unrolled;
+    //loop.vectorize(ni);
```
Comment (ZolotukhinM) [inline benchmarks/cpp/tensorexpr/bench_gemm.cpp:252]: "Vec" in the name is a lie :) Could you please add a TODO or some description why this doesn't work yet?

**Review by ZolotukhinM — APPROVED** [review]
Looks good to me as long as it builds and runs!

File: benchmarks/cpp/tensorexpr/bench_gemm.cpp [hunk 4]
```diff
@@ -127,6 +127,216 @@ BENCHMARK_DEFINE_F(Gemm, TensorExprTile32x32)(benchmark::State& state) {
...
+    loop.reorderAxis(ni, k);
+  }
+  // mo, no, mi, k, ni ->
+  // mo, no, k, mi, ni
+  {
+    auto const& loops = loop.getLoopStmtsFor(CT);
+    te::For* mi = loops[2];
+    te::For* k = loops[3];
+    loop.reorderAxis(mi, k);
+  }
```
Comment (navahgar) [inline benchmarks/cpp/tensorexpr/bench_gemm.cpp:182]: This code to tile in a specific way, 4x16 here, and reorder the loops appropriately, is being used repeatedly. How about creating a utility to tile the loops given the tile sizes?

**Review by navahgar — APPROVED** [review]
Mostly LGTM.

File: benchmarks/cpp/tensorexpr/bench_gemm.cpp (hunk 1 above)
Comment (bertmaher) [reply benchmarks/cpp/tensorexpr/bench_gemm.cpp:342]: oh this is actually a bug, there's a main.cpp now that adds BENCHMARK_MAIN

File: benchmarks/cpp/tensorexpr/bench_gemm.cpp (hunk 3 above)
Comment (bertmaher) [reply benchmarks/cpp/tensorexpr/bench_gemm.cpp:252]: @nickgg fixed this a while ago, I'll reenable it.

File: benchmarks/cpp/tensorexpr/bench_gemm.cpp (hunk 4 above)
Comment (bertmaher) [reply benchmarks/cpp/tensorexpr/bench_gemm.cpp:182]: I could factor something out here I guess, but I'd like us to improve the library API.  In tvm this whole giant pile of stuff is three lines:
```
mo, mi = s[C].split(m, 4)
no, ni = s[C].split(n, 16)
s[C].reorder(mo, no, k, mi, ni)
```

**Pull request closed without merging**